src/openclsim/critical_path/simulation_graph.py

```diff
@@ -64,36 +150,286 @@ def __check_dep(self, dependency_list):
+        together within the graph. Note that edges are added from the END of
+        the first activity to the START of the second.
+        """
+        for name_start, name_end in self.dependency_list:
+            # extract some information on the times
+            start_time = self.simulation_graph.nodes[name_start]["time"]
+            end_time = self.simulation_graph.nodes[name_end]["time"]
+            duration = end_time - start_time
+
+            if isinstance(duration, dt.timedelta):
+                duration = int(duration.total_seconds())
+            if not isinstance(duration, (float, int)):
+                raise TypeError(
+                    f"Duration computed as type {duration} " "is not supported!"
+                )
+            if round(duration, 4) != 0:
+                raise ValueError(
+                    f"dependency ({name_start}, {name_end})"
+                    f" with non zero duration ({duration}) not allowed!"
+                )
+            # add the edge from END to START
+            kwargs = {
+                "node_start": name_start,
+                "node_end": name_end,
+                "activity": None,
+                "start_time": start_time,
+                "state": None,
+                "duration": duration,
+                "end_time": end_time,
+                "cp_activity_id": None,
+                "edge_type": self.__EDGE_TYPES[1],
+                self.__COLNAME_CRITICAL: self.__CRITICAL[0],
+            }
+            self.simulation_graph.add_edge(name_start, name_end, **kwargs)
+
+    def __find_critical_edges(
+        self,
+        discount_graph=None,
+        list_critical=None,
+        list_noncritical=None,
+        to_discount=None,
+        max_duration=None,
+        t_max_end=None,
+        duration_discounted_prev=None,
+    ):
+        """
+        Create a list of edges on a longest path in the graph (recursively).
+
+        This method recursively builds a list of edges in the activity graph
+        which are on the longest path. In a nutshell the process is as follows:
+
+        0. Create a shadow/copy of the activity graph.
+        1. Find an initial longest path, mark all edges on it as such, and set
+           their 'duration' to 10**-4 on the shadow-graph (i.e. discount all
+           edges which have already been marked).
+        2. Find a new longest path in the shadow-graph, and check if the
+           duration in the original graph equals the maximum length. If so,
+           good! Mark all edges as critical and set duration in the shadow
+           graph to 10**-4. If not a feasible longest path, then mark the last
+           edge as noncritical, and set its weight in the shadow graph to
+           10**-8.
+        3. Repeat until all edges are marked as critical or non-critical.
+        """
+
+        feasible_longest = False  # all edges on path can be discounted at once
+
+        # the initial iteration
+        if discount_graph is None:
```

Comment:
naming reminds me a bit of supermarkets... Maybe we could choose another name here, but I have no alternative yet...

Reply:
haha..Yes maybe too much supermarket :)
changed to marked_edges_graph